jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,356 @@
+from jaxopt._src.tree_util import tree_l2_norm, tree_inf_norm, tree_sub, tree_add, tree_mul
+from jax.experimental import host_callback as hcb
+
+
+class LevenbergMarquardtState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  damping_factor: float
+  increase_factor: float
+  residual: Any
+  loss: Any
+  delta: Any
+  error: float
+  dparams_error: float
+  grad_error: float
+  gradient: Any
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class LevenbergMarquardt(base.IterativeSolver):
+  """Levenberg-Marquardt nonlinear least-squares solver.
+
+    Given the residual function `func` (x): R^m -> R^n, `least_squares` finds a
+    local minimum of the cost function F(x):
+
+    ```
+    argmin_x F(x) = 0.5 * sum(f_i(x)**2), i = 0, ..., m - 1
+    f(x) = func(x, *args)
+    ```
+
+    The convergence is achieved once the coeff update satisfies the criteria:
+    ```
+    ||dcoeffs||_2 <= xtol * (||coeffs||_2 + xtol)
+    ```
+    or the gradient satisfies ``||grad(f)||_inf <= gtol``.
+
+  Attributes:
+    residual_fun: a smooth function of the form ``residual_fun(x, *args,
+      **kwargs)``.
+    maxiter: maximum increase_factormber of iterations.
+    tol: tolerance.
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled
+      iterations.
+    implicit_diff_solve: the linear system solver to use.
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
+
+  Reference: This algorithm is for finding the best fit parameters based on the
+    algorithm 6.18 provided by K. Madsen & H. B. Nielsen in the book
+    "Introduction to Optimization and Data Fitting".
+  """
+  residual_fun: Callable
+
+  maxiter: int = 30
+  tol: float = 0.0
+
+  damping_parameter: float = 1e-6
+  xtol: float = 1e-8
+  gtol: float = 1e-8
+  solver: Literal['cg', 'bicgstab', 'gmres', 'cholesky', 'inv'] = 'cholesky'
+
+  verbose: bool = False
+  implicit_diff: bool = True
+  implicit_diff_solve: Optional[Callable] = None
+  has_aux: bool = False
+  jit: base.AutoOrBoolean = 'auto'
+  unroll: base.AutoOrBoolean = 'auto'
+
+  def init_state(self, init_params: Any, *args,
+                 **kwargs) -> LevenbergMarquardtState:
+    """Initialize the solver state.
+
+    Args:
+      init_params: pytree containing the initial parameters.
+      *args: additional positional arguments to be passed to ``residual_fun``.
+      **kwargs: additional keyword arguments to be passed to ``residual_fun``.
+
+    Returns:
+      state
+    """
+    # Compute actual values of state variables at init_param
+    residual, aux = self._fun_with_aux(init_params, *args, **kwargs)
+    gradient = self._jt_op(init_params, residual, *args, **kwargs)
+
+    grad_diff = tree_sub(tree_inf_norm(gradient), self.gtol)
```

Comment:
`tree_inf_norm` returns a scalar value so normally you can write

```suggestion
    grad_diff = tree_inf_norm(gradient) - self.gtol
```

You have a bunch of those in the code.

Reply:
Done. thanks for catching.